ssl/s3_enc.c

```diff
@@ -399,11 +418,14 @@ int ssl3_change_cipher_state(SSL *s, int which)
 
     OPENSSL_cleanse(&(exp_key[0]), sizeof(exp_key));
     OPENSSL_cleanse(&(exp_iv[0]), sizeof(exp_iv));
-    EVP_MD_CTX_cleanup(&md);
     return (1);
  err:
     SSLerr(SSL_F_SSL3_CHANGE_CIPHER_STATE, ERR_R_MALLOC_FAILURE);
  err2:
+    if (s->enc_write_ctx && !(which & SSL3_CC_READ)) {
+        EVP_CIPHER_CTX_free(s->enc_write_ctx);
+        s->enc_write_ctx = NULL;
+    }
```

Comment:
The only issue with enc_read_ctx I encountered was the one with aes_gcm_cleanup,
but I agree, that it would not hurt to do the same to enc_read_ctx as well. 

Yes, the application is not aware that the SSL socket can get into that kind of
dangerous state, and does always try to call SSL_shutdown.

As far as I remember SSL_shutdown returns once an error SSL_ERROR_SSL,
and then has to be called again to start the bidirectional shutdown phase.